x/aswap/handler.go

```diff
@@ -0,0 +1,292 @@
+		return nil, err
+	}
+	// return id of swap to use in future calls
+	res := &weave.DeliverResult{
+		Data: obj.Key(),
+	}
+	return res, h.bucket.Save(db, obj)
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h CreateSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateSwapMsg, error) {
+	var msg CreateSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	if IsExpired(ctx, msg.Timeout) {
+		return nil, errors.Wrap(errors.ErrInvalidInput, "timeout in the past")
+	}
+
+	// Sender must authorize this
+	if !h.auth.HasAddress(ctx, msg.Src) {
+		return nil, errors.ErrUnauthorized
+	}
+
+	// Leave the most expensive operation till we've sanity-checked everything else.
+	_, err := loadSwap(h.bucket, db, msg.PreimageHash)
+	switch {
+	case !errors.ErrEmpty.Is(err):
+		return nil, err
+	case err == nil:
+		return nil, errors.Wrap(errors.ErrDuplicate, "swap with the same preimage")
+	}
+
+	return &msg, nil
+}
+
+// ReleaseSwapHandler releases the amount to recipient.
+type ReleaseSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.Controller
+}
+
+var _ weave.Handler = ReleaseSwapHandler{}
+
+// Check just verifies it is properly formed and returns
+// the cost of executing it
+func (h ReleaseSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &weave.CheckResult{GasAllocated: releaseSwapCost}, nil
+}
+
+// Deliver moves the tokens from swap account to the receiver if
+// all preconditions are met. When the swap account is empty it is deleted.
+func (h ReleaseSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	swap, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	swapAddr := weave.NewAddress(swap.PreimageHash)
+	amount, err := h.bank.Balance(db, swapAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// withdraw the money from swap to recipient
+	if err := moveCoins(db, h.bank, swapAddr, swap.Recipient, amount); err != nil {
+		return nil, err
+	}
+
+	// Delete swap when empty.
+	if err := h.bucket.Delete(db, swap.PreimageHash); err != nil {
+		return nil, err
+	}
+	return &weave.DeliverResult{}, nil
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h ReleaseSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*Swap, error) {
+	var msg ReleaseSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	swap, err := loadSwap(h.bucket, db, hashPreimage(msg.Preimage))
+	if err != nil {
+		return nil, err
+	}
+
+	// Sender must authorize this.
+	if !h.auth.HasAddress(ctx, swap.Src) {
+		return nil, errors.ErrUnauthorized
+	}
+
+	if IsExpired(ctx, swap.Timeout) {
```

Comment:
req: I don't think we should enforce the claim before swap expiration. It is in the best interest of the receiver to sent the release msg before the timeout. Otherwise the original swap creator could claim via returnMsg himself.